File: tools/registry.go
package tools

import (
	"context"
	"fmt"
	"strings"

	"github.com/user/uclaw/provider"
)

type ToolRegistry struct {
	tools map[string]Tool
}

func NewToolRegistry() *ToolRegistry {
	return &ToolRegistry{tools: make(map[string]Tool)}
}

func (r *ToolRegistry) Register(t Tool) {
	r.tools[t.Name()] = t
}

func (r *ToolRegistry) Unregister(name string) {
	delete(r.tools, name)
}

func (r *ToolRegistry) Get(name string) Tool {
	return r.tools[name]
}

func (r *ToolRegistry) Has(name string) bool {
	_, ok := r.tools[name]
	return ok
}

func (r *ToolRegistry) GetDefinitions(visibleOnly bool) []provider.ToolSchema {
	var defs []provider.ToolSchema
	for _, t := range r.tools {
		if visibleOnly && t.IsHidden() {
			continue
		}
		defs = append(defs, ToSchema(t))
	}
	return defs
}

func (r *ToolRegistry) Execute(ctx context.Context, name string, params map[string]interface{}) (*ToolResult, error) {
	hint := "\n\n[Analyze the error above and try a different approach.]"
	t := r.tools[name]
	if t == nil {
		names := make([]string, 0, len(r.tools))
		for n := range r.tools {
			names = append(names, n)
		}
		return nil, fmt.Errorf("tool '%s' not found. Available: %s%s", name, strings.Join(names, ", "), hint)
	}
	result, err := t.Execute(ctx, params)
	if err != nil {
		return NewTextResult(fmt.Sprintf("Error executing %s: %s%s", name, err, hint)), nil
	}
	if result.Text != "" && strings.HasPrefix(result.Text, "Error") {
		return NewTextResult(result.Text + hint), nil
	}
	return result, nil
}

func (r *ToolRegistry) ToolNames() []string {
	names := make([]string, 0, len(r.tools))
	for n := range r.tools {
		names = append(names, n)
	}
	return names
}

File: v1/tools/registry.py
"""Tool registry for dynamic tool management."""

from typing import Any

from .base import Tool


class ToolRegistry:
    """
    Registry for agent tools.

    Allows dynamic registration and execution of tools.
    """

    def __init__(self):
        self._tools: dict[str, Tool] = {}

    def register(self, tool: Tool) -> None:
        """Register a tool."""
        self._tools[tool.name] = tool

    def unregister(self, name: str) -> None:
        """Unregister a tool by name."""
        self._tools.pop(name, None)

    def get(self, name: str) -> Tool | None:
        """Get a tool by name."""
        return self._tools.get(name)

    def has(self, name: str) -> bool:
        """Check if a tool is registered."""
        return name in self._tools

    def get_definitions(self, visible_only: bool = False) -> list[dict[str, Any]]:
        """Get all tool definitions in OpenAI format.
        
        Args:
            visible_only: 如果为 True，只返回 hidden=False 的工具（用于 LLM 指令注入）
        """
        tools = self._tools.values()
        if visible_only:
            tools = [t for t in tools if not getattr(t, 'hidden', False)]
        return [tool.to_schema() for tool in tools]

    async def execute(self, name: str, params: dict[str, Any]) -> Any:
        """Execute a tool by name with given parameters."""
        _HINT = "\n\n[Analyze the error above and try a different approach.]"

        tool = self._tools.get(name)
        if not tool:
            return f"Error: Tool '{name}' not found. Available: {', '.join(self.tool_names)}"

        try:
            # Attempt to cast parameters to match schema types
            params = tool.cast_params(params)
            
            # Validate parameters
            errors = tool.validate_params(params)
            if errors:
                return f"Error: Invalid parameters for tool '{name}': " + "; ".join(errors) + _HINT
            result = await tool.execute(**params)
            if isinstance(result, str) and result.startswith("Error"):
                return result + _HINT
            return result
        except Exception as e:
            return f"Error executing {name}: {str(e)}" + _HINT

    @property
    def tool_names(self) -> list[str]:
        """Get list of registered tool names."""
        return list(self._tools.keys())

    def __len__(self) -> int:
        return len(self._tools)

    def __contains__(self, name: str) -> bool:
        return name in self._tools
